clang/lib/CodeGen/CGCall.cpp

```diff
@@ -4148,6 +4152,30 @@ static void emitWriteback(CodeGenFunction &CGF,
   assert(!isProvablyNull(srcAddr.getBasePointer()) &&
          "shouldn't have writeback for provably null argument");
 
+  if (CGF.getLangOpts().HLSL) {
+    if (!isa<OpaqueValueExpr>(writeback.CastExpr)) {
+      RValue TmpVal = CGF.EmitAnyExprToTemp(writeback.CastExpr);
+      if (TmpVal.isScalar())
+        CGF.EmitStoreThroughLValue(TmpVal, srcLV);
+      else {
+        llvm::Value *Val = CGF.Builder.CreateLoad(TmpVal.getAggregateAddress());
```

Comment:
There's a really awful type of lvalue cast that HLSL supports (which I haven't yet added to Clang). We allow something like this:

```hlsl
struct Dog {
  int legs[4];
};

struct Cat {
  int4 legs;
};

void pet(inout Cat C) {
  C.legs.x = 1;
}

export void call() {
  Dog d;
  pet((Cat)d);
}
```

In HLSL 2018, this conversion was allowed to be implicit, but we're slowly tightening up the rules.

I will pull this code out from this PR though, because it isn't possible yet to construct an expression that results in that write back formation.

Reply:
Well, that's awful, but it's supportable.  The easiest way to get out of the business of having to handle all these different evaluation kinds is to just change the AST to the representation you'll need to use to support non-POD types, because then the load/assignment will just be the normal code-gen for emitting assignments.